clash-prelude/src/Clash/Signal/Trace.hs

```diff
@@ -449,12 +494,70 @@ dumpVCD
   => (Int, Int)
   -- ^ (offset, number of samples)
   -> Signal domain a
-  -- ^ (One of) the output the circuit containing the traces
+  -- ^ (One of) the outputs of the circuit containing the traces
   -> [String]
   -- ^ The names of the traces you definitely want to be dumped in the VCD file
   -> IO (Either String Text.Text)
 dumpVCD = dumpVCD# traceMap#
 
+-- | Dump a number of samples to a replayable bytestring.
+dumpReplayable
+  :: forall a domain
+   . Undefined a
+  => Int
+  -- ^ Number of samples
+  -> Signal domain a
+  -- ^ (One of) the outputs of the circuit containing the traces
+  -> String
+  -- ^ Name of trace to dump
+  -> IO ByteString
+dumpReplayable n oSignal traceName = do
+  waitForTraces# traceMap# oSignal [traceName]
+  replaySignal <- (Map.! traceName) <$> readIORef traceMap#
+  let (tRep, _period, _width, samples) = replaySignal
+  pure (ByteStringLazy.concat (tRep : map encode (take n samples)))
+
+-- | Take a serialized signal (dumped with @dumpReplayable@) and convert it
+-- back into a signal. Will error if dumped type does not match requested
+-- type. The first value in the signal that fails to decode will stop the
+-- decoding process and yield an error. Not that this always happens if you
+-- evaluate more values than were originally dumped.
+replay
+  :: forall a domain n
+   . Typeable a
+  => Undefined a
+  => BitPack a
+  => n ~ BitSize a
+  => KnownNat n
+  => ByteString
+  -> Either String (Signal domain a)
```

Comment:
```suggestion
  :: forall a domain n
   . (Typeable a, Undefined a, BitPack a, n ~ BitSize a, KnownNat a)
  => ByteString
  -> Either String (Signal domain a)
```

Reply:
I much prefer to do it on multiple lines, because there are so many constraints (I think my mental/visual limit is around ~3). I'll keep the "tuple" notation of course, as we discussed for [the unofficial official style guide](https://github.com/clash-lang/clash-compiler/pull/584).